include/treelite/tree_impl.h

```diff
@@ -737,24 +742,12 @@ template <typename ThresholdType, typename LeafOutputType>
 inline void
 Tree<ThresholdType, LeafOutputType>::SetLeafVector(
     int nid, const std::vector<LeafOutputType>& node_leaf_vector) {
-  const std::size_t end_oft = leaf_vector_offset_.Back();
-  const std::size_t new_end_oft = end_oft + node_leaf_vector.size();
-  if (end_oft != leaf_vector_.Size()) {
-    throw std::runtime_error("Invariant violated");
-  }
-  if (!std::all_of(&leaf_vector_offset_.at(nid + 1), leaf_vector_offset_.End(),
-                   [end_oft](std::size_t x) { return (x == end_oft); })) {
-    throw std::runtime_error("Invariant violated");
-  }
-  // Hopefully we won't have to move any element as we add leaf vector elements for node nid
+  std::size_t begin = leaf_vector_.Size();
+  std::size_t end = begin + node_leaf_vector.size();
   leaf_vector_.Extend(node_leaf_vector);
-  if (new_end_oft != leaf_vector_.Size()) {
-    throw std::runtime_error("Invariant violated");
-  }
-  std::for_each(&leaf_vector_offset_.at(nid + 1), leaf_vector_offset_.End(),
-                [new_end_oft](std::size_t& x) { x = new_end_oft; });
-
-  Node& node = nodes_.at(nid);
+  leaf_vector_begin_[nid] = begin;
+  leaf_vector_end_[nid] = end;
+  Node &node = nodes_.at(nid);
```

Comment:
It might be prudent to throw an error if the user attempts to call `SetLeafVector` out-of-order, e.g. `SetLeafVector(nid=1)` followed by `SetLeafVector(nid=0)`. The previous version of the code performs this check. Do you think adding the check will greatly impact performance?

In general, we should ensure that the user initializes nodes in the order of node IDs: node 0, node 1, node 2, and so forth.

Reply:
I don't think there is actually any problem with setting this out of order in the current design. We have a unique map for each nid to the leaf vector, and they are independent, so I don't think the ordering of the leaf vector matters, unless I have missed something. I think the previous throw statement was one of the performance issues unfortunately.

It looks like your python linter has updated and is failing the unchanged python code. What would you like me to do here?

Reply:
@RAMitchell Actually you are right. As long as you don't call `SetLeafVector` twice for the same node, things will be okay. How about if we add a check to ensure that `SetLeafVector` gets called at most once for each node ID?

Reply:
I think this will still not result in any critical failure, there will just be some unused memory in the leaf vector. I can do this change if you want, but I should redo the benchmarks to check if the throw statement kills performance again.

Reply:
Got it. No need to re-run the benchmarks. Let's keep the current logic as it is.